Implementation

lennym opened the pull request:
Opening as a PR for easy review of ongoing development and feedback.

Feel free to merge at any point as it doesn't go anywhere if it is merged.
# TODO
- [x] Documentation
- [x] Tests
- [x] Implement stream flattening - i.e. hooking into `done` event to provide a simple callback interface as well as event-driven result streams
- [ ] Implement plugin system to allow adding additional custom methods onto SDK

lennym committed:
First pass implementation

lennym committed:
Remove auto send of join event to sockets

Instead of sending a separate `join` signal to the server we can simply auto-join the room when the server receives a query, since there is no context in which it will make sense for a `query` to not be accompanied by a `join`

lennym committed:
Send package version with queries

This will allow the receiving API server to ensure it is able to support the client version, and can send an appropriate message when it is not able to comply.

lennym committed:
Use globally configured timeout if it has been set

lennym committed:
Add a webpack watch command

lennym committed:
Move all dependencies to dev dependencies

Only the webpack compiled code is actually needed by installers, so there's no need for anything to be a prod dependency.

lennym committed:
Move sandbox to ./sandbox and document

lennym committed:
Fix bug with undefined timeout

lennym committed:
npm ignoe all the things

lennym committed:
Documentation

lennym committed:
Update comments in generated primus script

lennym committed:
Fix default url option in README

lennym committed:
Simplify socket module to remove one line, single call function

lennym commented on test/query.test.js:
```diff
@@ -0,0 +1,10 @@
+  it('', () => {
+    const out = query();
+
+  });
```
What do you mean "probably needs some more tests"??

nelsonic replied:
@lennym 
![lgtm-stevie-wonder](https://cdn.meme.am/instances/500x/67132980.jpg)
😆

lennym committed:
Remove copied comments from webpack config

lennym committed:
Support browser and node versions

The only difference required to support browser and node versions is the Primus socket library we use. By passing a Socket constructor in as a configuration option to the main SDK we can support different entry points for browser and node versions.

lennym committed:
Add stub backend for use with unit tests

Backend can be configured to provide a particular set of responses at particular intervals

lennym committed:
Add tests

lennym committed:
Add tests for concurrent queries

lennym commented:
Documentation and Tests now marked as "Done" on the grounds that there literally are some.

lennym committed:
Remove  from TODO

lennym committed:
Add docs note re redis

lennym committed:
Add `result` method to query to flatten stream into an array of results

nelsonic commented:
Hi @lennym code looks good _so far_.
I decided to run the tests, checkout out the `implement` branch and did `npm install && npm t` got:
![missing-engine-io-client-module](https://cloud.githubusercontent.com/assets/194400/18210183/3bce3baa-712f-11e6-8777-c6b34c217505.png)

is the `engine.io-client` _assumed_ or will it be downloaded from the `sdk-server` ?

When I did `npm install --save engine.io-client` and _then_ `npm test` the tests _all passed_:
![tests-pass-with-engine io-client](https://cloud.githubusercontent.com/assets/194400/18210276/afa74b20-712f-11e6-90a0-3fd757a67dc7.png)

nelsonic committed:
adds engine.io-client devDependency see: https://github.com/numo-labs/sdk/pull/1#issuecomment-244418261

nelsonic committed:
remove .DS_Store !

nelsonic commented:
Looking forward to seeing the plugin system. 😉

nelsonic commented:
@lennym have you had a look at http://rollupjs.org/ as alternative to WebPack ?

lennym commented:
@nelsonic I did not consider any other compilers, on the basis that webpack performs the job we need more than adequately, and researching and analysing potential replacements for already working components is a recipe for not getting anything actually delivered. 

If it transpires in future that webpack is somehow lacking, or rollup (or browserify for that matter) are considered overwhelmingly superior for whatever reason, then since the code is simply common js modules, it will be trivially easy to replace one with the other.

nelsonic commented:
@lennym agree. 👍

lennym committed:
Remove event listeners and unsubscribe channel on error

nelsonic commented on test/helpers/index.js:
```diff
@@ -0,0 +1,2 @@
+const chai = require('chai');
+global.expect = chai.expect;
```
@lennym does this _need_ to be a _Global_...?

lennym replied:
No, the alternative is to put those two lines in the top of every test file, but given that test suites are already littered with globals (`describe`, `it`, `before`, etc) then adding one more is a cleaner approach that repeating that same code block all over the place.

lennym committed:
Point default backend to new EB deployed instance

lennym committed:
Update sample queries in sandbox page

lennym committed:
Refactor tests to use a mock websocket server instead of actual backend

The actual backend has been refactored to integrate with external lambda functions so cannot be configured on the fly in a way compatible with using it for unit tests here.

Instead use a mock websocket server to imitate the output of the backend.

lennym committed:
Add documentation for `entities` parameter

lennym committed:
Add `lambda` config option

Use a configuration option of `lambda` to define query distributor lambda function name.Primarily to support alternative lambda options in development.

lennym committed:
Add support for a loglevel parameter
